docs/Cursors.md

```diff
@@ -0,0 +1,287 @@
+   <- c.after()
+```
+
+## Other Cursor Navigation
+
+- `GapCursor.anchor()`: Returns cursor to the statement this gap is anchored to
+
+- `BlockCursor.expand(delta_lo=None, delta_hi=None)`: Returns an expanded block cursor 
+   - `delta_lo`/`delta_hi` specify statements to add at start/end; `None` means expand fully
+   - Ex. in `s1; s2; s3`, if `c` is a `BlockCursor` pointing `s1; s2`, then `c.expand(0, 1)` returns a new `BlockCursor` pointing `s1; s2; s3`
+- `BlockCursor.before()`: Returns `GapCursor` before block's first statement
+- `BlockCursor.after()`: Returns `GapCursor` after block's last statement
+
+## Cursor inspection
+
+`StmtCursor`s wrap the underlying Exo IR object and can be inspected.
+   - Ex. check cursor type with `isinstance(c, PC.AllocCursor)`
+
+`StmtCursor`s are one of the following types.
+
+#### `ArgCursor`
+
+Represents a cursor pointing to a procedure argument of the form:
+```
+name : type @ mem
+```
+
+Methods:
+- `name() -> str`: Returns the name of the argument.
+- `mem() -> Memory`: Returns the memory location of the argument.
+- `is_tensor() -> bool`: Checks if the argument is a tensor.
+- `shape() -> ExprListCursor`: Returns a cursor to the shape expression list.
+- `type() -> API.ExoType`: Returns the type of the argument.
+
+#### `AssignCursor`
+
+Represents a cursor pointing to an assignment statement of the form:
+```
+name[idx] = rhs
+```
+
+Methods:
+- `name() -> str`: Returns the name of the variable being assigned to.
+- `idx() -> ExprListCursor`: Returns a cursor to the index expression list.
+- `rhs() -> ExprCursor`: Returns a cursor to the right-hand side expression.
+- `type() -> API.ExoType`: Returns the type of the assignment.
+
+#### `ReduceCursor`
+
+Represents a cursor pointing to a reduction statement of the form:
+```
+name[idx] += rhs
+```
+
+Methods:
+- `name() -> str`: Returns the name of the variable being reduced.
+- `idx() -> ExprListCursor`: Returns a cursor to the index expression list.
+- `rhs() -> ExprCursor`: Returns a cursor to the right-hand side expression.
+
+
+#### `AssignConfigCursor`
+
+Represents a cursor pointing to a configuration assignment statement of the form:
+```
+config.field = rhs
+```
+
+Methods:
+- `config() -> Config`: Returns the configuration object.
+- `field() -> str`: Returns the name of the configuration field being assigned to.
+- `rhs() -> ExprCursor`: Returns a cursor to the right-hand side expression.
+
+#### `PassCursor`
+
+Represents a cursor pointing to a no-op statement:
+```
+pass
+```
+
+#### `IfCursor`
+
+Represents a cursor pointing to an if statement of the form:
+```
+if condition:
+    body
+```
+or
+```
+if condition:
+    body
+else:
+    orelse
+```
+Returns an invalid cursor if `orelse` isn't present.
+
+Methods:
+- `cond() -> ExprCursor`: Returns a cursor to the if condition expression.
+- `body() -> BlockCursor`: Returns a cursor to the if body block.
+- `orelse() -> Cursor`: Returns a cursor to the else block (if present).
+
```

Comment:
`orelse() -> BlockCursor | InvalidCursor`: Returns a cursor to the else block if present, otherwise returns an invalid cursor.